### PR TITLE
Stop when declaring an already existing task or a loop

### PR DESCRIPTION
**Pull request by xavierpriour**
This patch detects 2 error cases:
- registering a task under the name of an already existing one:

```
grunt.loadNpmTasks('grunt-contrib-uglify');
grunt.registerTask('uglify', ['a', 'b']);
```
- registering a looping task:

```
grunt.registerTask('build', ['a', 'build', 'b']);
```

It then throws an exception with an (hopefully useful) error message.

This fixes my issue #1342 and replaces FAQ http://gruntjs.com/frequently-asked-questions#why-am-i-getting-a-maximum-call-stack-size-exceeded-error.

### CODE REVIEW

**Comment by xavierpriour** [comment]
@cowboy @tkellen @vladikoff - does this PR need anything done to get merged?
Thanks

**Comment by silkentrance** [comment]
I do like the feature that one can redefine existing tasks and call upon the original task when necessary, so perhaps a simple warning instead of throwing an error would be less breaking with existing gruntfiles.

**Comment by silkentrance** [comment]
As for the recursion thing. Please consider dependent tasks, that are also multi tasks and include a dependency towards a requiring task, e.g.

```
grunt.registerTask("uglify", ["a", "b"]);
grunt.registerTask("a", ["uglify", "c"]);
```

I think that by building a dependency graph would solve the trick, e.g. a simple hash would do

```
dependencyGraph = {
   "uglify" : {
        "requires" : ["a", "b"],
        "requiredby" : ["a"]
   },
   "a" : {
       "requires" : ["uglify", "c"],
       "requiredby" : ["uglify"]
   }
}
```

And, when all tasks have been registered, simply searching for duplicates in requires and requiredby would reveal all recursively defined tasks.

Of course, this still does not cover the case where these loops are defined over multiple tasks, e.g.

```
grunt.registerTask("uglify", ["a", "b"]);
grunt.registerTask("a", ["c"]);
grunt.registerTask("c", ["uglify"]);
```

which would then create a dependency graph something like

```
dependencyGraph = {
   "uglify" : {
        "requires" : ["a", "b"],
        "requiredby" : ["c"]
   },
   "a" : {
       "requires" : ["c"],
       "requiredby" : []
   }
   "c" : {
       "requires" : ["uglify"],
       "requiredby" : ["a"]
   }
}
```

**Comment by silkentrance** [comment]
@xavierpriour please have a look, i commented on your PR.

**Comment by xavierpriour** [comment]
Hi @silkentrance thanks for your comments.

- regarding "I do like the feature that one can redefine existing tasks and call upon the original task when necessary", currently grunt replaces the existing task with the new one, so there is no way to call upon the original task. Proper way is to rename the existing then register the new one.
- regarding task cycles, detecting cycles is a complex task, and grunt does not currently expose enough information internally (for example, it does not store the list of tasks of a multiTask) to go beyond my simplistic case.

So basically I'm aiming for the 20/80 (20% of work delivering 80% of benefits) in this PR :)

**Comment by silkentrance** [comment]
@xavierpriour Well, it actually does expose the required information.
See the parent task with which all tasks will be registered with? It holds a map of all defined tasks ```_tasks```, which are objects 

```
{ name : String, info : String, fn : Function|Array<String> }
```

From this, the dependency map could be compiled from and would then be checked prior to that a task is actually being run...

**Comment by silkentrance** [comment]
@xavierpriour The Task class should actually be refactored: it serves as a registry for all tasks and also as a factory for the actual task instances and as the task instance that can then be run...
SOC anyone?

**Comment by silkentrance** [comment]
@xavierpriour As for the other use case where one would overwrite existing other tasks, I could think of a use case where I´d have a custom tasks, say ´uglify´ which would first require babel to be run in order to compile my sources and then I would have a second task that would then register the npm tasks from grunt-contrib-uglify and will then run ´uglify´ again.

However, this would compete with the cycle detection, so yes, bailing out on the duplicate task seems to be just fine :grin:.

**Comment by xavierpriour** [comment]
@silkentrance re: grunt exposing the info, it seems to me that currently, for a multiTask, the `fn` property holds a function, not the list of strings (tasks.js line 67: `fn = this.run.bind(this, fn);`), so no luck here :(

I agree refactoring Task would be a great idea (a richer object would enable more powerful plugins), but you might want to onboard grunt core members before starting...

**Comment by silkentrance** [comment]
@xavierpriour shite, I overlooked that one. Nice catch. How about adding the missing information then to the task object, e.g.

```
fn_orig = null;
[...]
fn_orig = fn;
fn = ...bind...();
[...]
this._tasks[name] = { name : name, info : info, fn : fn, queue : fn_orig };
```

From queue one could then retrieve the names of the tasks...

**Comment by xavierpriour** [comment]
Obsolete I guess...
